src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -26,10 +29,23 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler
     [ExportLspMethod(LSP.Methods.TextDocumentCompletionName)]
     internal class CompletionHandler : AbstractRequestHandler<LSP.CompletionParams, LSP.CompletionItem[]>
     {
+        private readonly ImmutableHashSet<string> _csTriggerCharacters;
+        private readonly ImmutableHashSet<string> _vbTriggerCharacters;
+
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-        public CompletionHandler(ILspSolutionProvider solutionProvider) : base(solutionProvider)
+        public CompletionHandler(
+            ILspSolutionProvider solutionProvider,
+            [ImportMany] IEnumerable<Lazy<CompletionProvider, CompletionProviderMetadata>> completionProviders)
+            : base(solutionProvider)
         {
+            var csCompletionProviders = completionProviders?.Where(lz => lz.Metadata.Language == LanguageNames.CSharp).ToImmutableArray();
+            var vbCompletionProviders = completionProviders?.Where(lz => lz.Metadata.Language == LanguageNames.VisualBasic).ToImmutableArray();
+
+            _csTriggerCharacters = csCompletionProviders.HasValue ? csCompletionProviders.Value.SelectMany(
+                lz => GetTriggerCharacters(lz.Value)).Distinct().Select(c => c.ToString()).ToImmutableHashSet() : ImmutableHashSet.Create<string>();
```

Comment:
I *think* you don't need to do distinct here, ToImmutable should just add the unique (but check to make sure)